Runtime exception on permission request

I'm using flutter permission handler package to request permissions in different parts of the app and since I've added this package there seems to be problems in requesting any kind of permissions!
If I want to get camera perm for example in another route totally separate from this package's parts it would cause error and explicitly mentions OnAudioQueryPlugin.kt:157

setting permission request to false or null in querysongs didn't solve the problem(using only built in permission request functionality is fine for getting storage perm but permission handler problem still exists for other parts)

`D/AndroidRuntime(21268): Shutting down VM
E/AndroidRuntime(21268): FATAL EXCEPTION: main
E/AndroidRuntime(21268): Process: com.example.rgb, PID: 21268
E/AndroidRuntime(21268): java.lang.RuntimeException: Failure delivering result ResultInfo{who=@android:requestPermissions:, request=24, result=-1, data=Intent { act=android.content.pm.action.REQUEST_PERMISSIONS (has extras) }} to activity {com.example.rgb/com.example.rgb.MainActivity}: java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime(21268): 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4360)
E/AndroidRuntime(21268): 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:4402)
E/AndroidRuntime(21268): 	at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:49)
E/AndroidRuntime(21268): 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
E/AndroidRuntime(21268): 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
E/AndroidRuntime(21268): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
E/AndroidRuntime(21268): 	at android.os.Handler.dispatchMessage(Handler.java:106)
E/AndroidRuntime(21268): 	at android.os.Looper.loop(Looper.java:193)
E/AndroidRuntime(21268): 	at android.app.ActivityThread.main(ActivityThread.java:6669)
E/AndroidRuntime(21268): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(21268): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/AndroidRuntime(21268): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
E/AndroidRuntime(21268): Caused by: java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime(21268): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:155)
E/AndroidRuntime(21268): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.error(MethodChannel.java:243)
E/AndroidRuntime(21268): 	at com.lucasjosino.on_audio_query.OnAudioQueryPlugin.onRequestPermissionsResult(OnAudioQueryPlugin.kt:157)
E/AndroidRuntime(21268): 	at io.flutter.embedding.engine.FlutterEngineConnectionRegistry$FlutterEngineActivityPluginBinding.onRequestPermissionsResult(FlutterEngineConnectionRegistry.java:712)
E/AndroidRuntime(21268): 	at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.onRequestPermissionsResult(FlutterEngineConnectionRegistry.java:412)
E/AndroidRuntime(21268): 	at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onRequestPermissionsResult(FlutterActivityAndFragmentDelegate.java:627)
E/AndroidRuntime(21268): 	at io.flutter.embedding.android.FlutterActivity.onRequestPermissionsResult(FlutterActivity.java:640)
E/AndroidRuntime(21268): 	at android.app.Activity.dispatchRequestPermissionsResult(Activity.java:7608)
`


Flutter 2.0.4 stable channel
on_audio_query: ^1.0.6
permission_handler: ^6.1.3

compileSdkVersion 28
minSdkVersion 24
targetSdkVersion 29

So, just seeing the error log I noticed that this error it's a very strange and specific.

First:

> Failure delivering result ResultInfo{who=android:requestPermissions:, request=24, result=-1, data=Intent { act=android.content.pm.action.REQUEST_PERMISSIONS (has extras) }} to activity {com.example.rgb/com.example.rgb.MainActivity}: java.lang.IllegalStateException: Reply already submitted

When calling a request permission will return a code with the result, every request needs a unique code:

> request=24

None of my requests codes has 24, looking at permission_handler plugin you will find:

> static final int PERMISSION_CODE = 24;

And:

> Caused by: java.lang.IllegalStateException: Reply already submitted E/AndroidRuntime(21268): at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:155)

This means that flutter already received the result. Everytime you "invoke" a method from flutter to native code will return a result with some content or a error.

Here the strange part:

> E/AndroidRuntime(21268): at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.error(MethodChannel.java:243) E/AndroidRuntime(21268): at com.lucasjosino.on_audio_query.OnAudioQueryPlugin.onRequestPermissionsResult(OnAudioQueryPlugin.kt:157)

When android receive the request code and this code does't exist will return this:

(OnAudioQueryPlugin.kt:157):

> else -> pResult.error(channelError, "RequestCode: [$requestCode] don't exist, probably my fault, open a issue in GitHub", null)

**So... I don't now how or why this are happing, but, looks like when calling some permission from permission_handler, my plugin receive this code and after checking the code and realizing that don't exist, throw an error and return to flutter.**

And...

> setting permission request to false or null in querysongs didn't solve the problem

When setting to false my plugin will totaly ignore the permission. So, i don't really now :/

> my plugin recive this code and after checking the code and realizing that don't exist

Communications between flutter plugins using native code should't exist, idk how this is possible.

--- a/android/src/main/kotlin/com/lucasjosino/on_audio_query/OnAudioQueryPlugin.kt
+++ b/android/src/main/kotlin/com/lucasjosino/on_audio_query/OnAudioQueryPlugin.kt
@@ -149,17 +149,39 @@ class OnAudioQueryPlugin: FlutterPlugin, MethodCallHandler, ActivityAware,
   }
 
   override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>?, grantResults: IntArray?): Boolean {
+    // When `pResult` is not initialized the permission request did not originate from the
+    // on_audio_query plugin, so return `false` to indicate the on_audio_query plugin is not
+    // handling the request result and Android should continue executing other registered handlers.
+    if (!this::pResult.isInitialized) {
+      return false
+    }
+
+    // When the incoming request code doesn't match the request codes defined by the on_audio_query
+    // plugin return `false` to indicate the on_audio_query plugin is not handling the request
+    // result and Android should continue executing other registered handlers.
+    if (requestCode != onRequestCode && requestCode != onExternalRequestCode) {
+      return false
+    }
+
     val isPermissionGranted = if (grantResults != null) grantResults.isNotEmpty()
             && grantResults[0] == PackageManager.PERMISSION_GRANTED else false
+
     when (requestCode) {
-      onRequestCode -> if (isPermissionGranted) onAudioController.onAudioController()
-      else if (retryRequest) onRetryRequestPermission()
-      //
-      onExternalRequestCode -> if (isPermissionGranted) pResult.success(true)
-      else if (retryRequest) onRetryRequestPermission() else pResult.success(false)
-      //
-      else -> pResult.error(channelError, "RequestCode: [$requestCode] don't exist, probably my fault, open a issue in GitHub", null)
+      onRequestCode -> {
+        if (isPermissionGranted) onAudioController.onAudioController()
+        else if (retryRequest) onRetryRequestPermission()
+      }
+      onExternalRequestCode -> {
+        when {
+            isPermissionGranted -> pResult.success(true)
+            retryRequest -> onRetryRequestPermission()
+            else -> pResult.success(false)
+        }
+      }
     }
-    return false
+
+    // Return `true` here to indicate that the on_audio_query plugin handled the permission request
+    // result and Android should not continue executing other registered handlers.
+    return true;
   }
 }

